plugins/permission-node/src/integration/createPermissionIntegration.test.ts

```diff
@@ -0,0 +1,293 @@
+          .send({
+            resourceRef: 'default:test/resource',
+            resourceType: 'test-resource',
+            conditions: {
+              anyOf: [],
+            },
+          });
+
+        expect(response.status).toEqual(200);
+        expect(response.body).toEqual({ result: AuthorizeResult.DENY });
+      });
+
+      it('returns 400 when called with incorrect resource type', async () => {
+        const response = await request(app)
+          .post('/permissions/apply-conditions')
+          .send({
+            resourceRef: 'default:test/resource',
+            resourceType: 'test-incorrect-resource',
+            conditions: {
+              anyOf: [],
+            },
+          });
+
+        expect(response.status).toEqual(400);
+        expect(response.error && response.error.text).toMatch(
+          /unexpected resource type: test-incorrect-resource/i,
+        );
+      });
+
+      it('returns 400 when resource is not found', async () => {
+        mockGetResource.mockReturnValueOnce(Promise.resolve(undefined));
+
+        const response = await request(app)
+          .post('/permissions/apply-conditions')
+          .send({
+            resourceRef: 'default:test/resource',
+            resourceType: 'test-resource',
+            conditions: {
+              not: {
+                rule: 'testRule1',
+                params: ['a', 1],
+              },
+            },
+          });
+
+        expect(response.status).toEqual(400);
+        expect(response.error && response.error.text).toMatch(
+          /resource for ref default:test\/resource not found/i,
+        );
+      });
+
+      it.each([
+        undefined,
+        {},
+        { resourceType: 'test-resource-type' },
+        { resourceRef: 'test/resource-ref' },
+        {
+          resourceType: 'test-resource-type',
+          resourceRef: 'test/resource-ref',
+        },
+        { conditions: { anyOf: [] } },
+      ])(`returns 400 for invalid input %#`, async input => {
+        const response = await request(app)
+          .post('/permissions/apply-conditions')
+          .send(input);
+
+        expect(response.status).toEqual(400);
+        expect(response.error && response.error.text).toMatch(
+          /invalid request body/i,
+        );
+      });
+    });
+  });
+
+  describe('toQuery', () => {
+    it('converts conditions to plugin-specific queries using rule toQuery methods', () => {
+      const { toQuery } = testIntegration();
+
+      expect(
+        toQuery({
+          anyOf: [
+            {
+              allOf: [
+                { rule: 'testRule1', params: ['a', 1] },
+                { rule: 'testRule2', params: [{ foo: 'bar' }] },
+              ],
+            },
+            {
+              not: { rule: 'testRule1', params: ['b', 2] },
+            },
+          ],
+        }),
+      ).toEqual({
+        anyOf: [
+          {
+            allOf: [
+              { query: 'testRule1', params: ['a', 1] },
```

Comment:
Nitpicking here, but this test would still pass if `toQuery` returned the conditions without calling each rule's `toQuery`. Might be better to implement some sort of transform in the mock permission integration's rules

Reply:
I'm not totally sure I'm interpreting your comment in the right way, but I think your implication is that the expected result here is identical to the input criteria. It's actually not - there's a subtle difference, it renames the "rule" property to "query". Looking at it again though this is super non-obvious, so maybe it makes sense to transform the condition into something more visibly different. Will do that now!

Reply:
Done in 691483ddc2!

Reply:
Ah yes you're right! Thanks for pushing the change anyway 😄